# 第一题：使用sklearn的GaussianNB、BernoulliNB、MultinomialNB完成spambase垃圾邮件分类任务

实验内容：
1. 使用GaussianNB、BernoulliNB、MultinomialNB完成spambase邮件分类
2. 计算各自十折交叉验证的精度、查准率、查全率、F1值
3. 根据精度、查准率、查全率、F1值的实际意义以及四个值的对比阐述三个算法在spambase中的表现对比

# 1. 读取数据集

In [1]:
import numpy as np
spambase = np.loadtxt('data/spambase/spambase.data', delimiter = ",")
spamx = spambase[:, :57]
spamy = spambase[:, 57]

# 2. 导入模型

In [2]:
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.model_selection import cross_val_predict

In [4]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score

# 3. 计算十折交叉验证下，GaussianNB、BernoulliNB、MultinomialNB的精度、查准率、查全率、F1值

In [5]:
# YOUR CODE HERE
model = GaussianNB()
prediction = cross_val_predict(model, spamx, spamy.astype('int'), cv = 10)

acc_spam = accuracy_score(spamy,prediction)
pre_spam = precision_score(spamy,prediction)
rec_spam = recall_score(spamy,prediction)
f1_spam = f1_score(spamy,prediction)

print(acc_spam,pre_spam,rec_spam,f1_spam)

0.8217778743751358 0.7004440855874041 0.9569773855488142 0.8088578088578089


In [6]:
model = BernoulliNB()
prediction = cross_val_predict(model, spamx, spamy.astype('int'), cv = 10)

acc_spam = accuracy_score(spamy,prediction)
pre_spam = precision_score(spamy,prediction)
rec_spam = recall_score(spamy,prediction)
f1_spam = f1_score(spamy,prediction)

print(acc_spam,pre_spam,rec_spam,f1_spam)

0.8839382742881983 0.8813357185450209 0.815223386651958 0.8469914040114614


In [7]:
model = MultinomialNB()
prediction = cross_val_predict(model, spamx, spamy.astype('int'), cv = 10)

acc_spam = accuracy_score(spamy,prediction)
pre_spam = precision_score(spamy,prediction)
rec_spam = recall_score(spamy,prediction)
f1_spam = f1_score(spamy,prediction)

print(acc_spam,pre_spam,rec_spam,f1_spam)

0.786350793305803 0.7323628219484882 0.7214561500275786 0.7268685746040567


###### 双击此处填写
算法|精度|查准率|查全率|F1值
-|-|-|-|-
GaussianNB|0.8217778743751358|0.7004440855874041|0.9569773855488142|0.8088578088578089
MultinomialNB|0.8839382742881983|0.8813357185450209|0.815223386651958|0.8469914040114614
BernoulliNB|0.786350793305803|0.7323628219484882|0.7214561500275786|0.7268685746040567

###### 双击此处填写分析

分析如下：GaussianNB的查全率最高，查准率最低，综合效果较好；
BernoulliNB的精度、查准率、F1值最高，综合效果最好；
MultinomialNB的精度、查全率、F1值均为最低，综合效果最差；